### PR TITLE
sys: add ChaCha stream cipher and PRNG

### PR DESCRIPTION
**Pull request by Kijewski**
This implementation is optimized for a little code and data size, not
for speed. IMO the code is more readable than in the reference
implementation.

The biggest advantage of ChaCha over other stream ciphers is the very
little data usage with only 64 bytes of context, and its good encryption
speed.

Also part of this PR is pseudo-random number generator, that just
returns the keystream of a randomly initialized ChaCha context.

### CODE REVIEW

File: sys/crypto/Makefile [hunk 1]
```diff
@@ -1 +1,7 @@
+ifeq (, ${RIOT_CHACHA_PRNG_DEFAULT})
+    RIOT_CHACHA_PRNG_DEFAULT := $(shell jot -w '0x%08xull,' -r 16 0 0x7ffffffe)
```
Comment (phiros) [inline sys/crypto/Makefile]: Please add the debian package name of ```jot``` to:
* dist/tools/travis-scripts/get-pkg-list.py
* dist/Dockerfile
* riotdocker repo (in a separate PR)

<offtopic>
We should probably create a proper list of the programs needed to build
RIOT in the very near future. Not everyone has tools like jot installed...</offtopic>

File: sys/crypto/chacha.c [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+ * Copyright (C) 2008  D. J. Bernstein  (dedicated to the public domain)
+ * Copyright (C) 2015  René Kijewski  <rene.kijewski@fu-berlin.de>
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
```
Comment (phiros) [inline sys/crypto/chacha.c]: There are a lot of trailing whitespaces in this file.
Check https://travis-ci.org/RIOT-OS/RIOT/jobs/67709906 for more details

File: sys/crypto/Makefile (hunk 1 above)
Comment (phiros) [inline sys/crypto/Makefile]: BTW: you could also check whether ```jot``` exists in the path. If it doesn't exist you can just set this variable to a default value. For testing purposes this should be fine.
However, this needs to be mentioned somewhere (Printing a warning message during compiling should suffice)

Comment (Kijewski) [inline sys/crypto/Makefile]: I switched to urandom and hexdump. I guess it's the right tool here anyway.

File: sys/crypto/chacha.c (hunk 2 above)
Comment (Kijewski) [inline sys/crypto/chacha.c]: Fixed

**Comment by rfswarm** [comment]
what you don't like about #2387? criticizing it (by blaming me I am a copycat) and re-implementing it - yeah ty - I don't want to rude but what mostly makes your code smaller is the use of memcpy vs U8TO32_LITTLE.

**Comment by phiros** [comment]
@OlegHahm I would favor @Kijewski pull request since I prefer the use of memcpy over U8TO32_LITTLE. What is your opinion on this? 

@rfswarm As far as I can see the argument for #3229 over #2387 is that this PR is easier to read
(which is IMHO much more important in a security context than speed).

**Comment by OlegHahm** [comment]
@phiros, I haven't compared the two PRs, but I second your arguments.

**Comment by phiros** [comment]
@OlegHahm OK then. 

@Kijewski PR seems fine to me + Travis is happy -> ACK and go

**Pull request merged**